### PR TITLE
898 - crash measure report for ALL patients

### LINKED ISSUE #898: Reports Page crashes for 'All' selection
**Describe the bug**

Reports is working fine for individual patients i.e. when we select 'Individual' and a patient. But if we keep 'All' it crashes due to NPE as PatientItem in generation method has a null check

**To Reproduce**
1. Login to 'anc' appId for ANC app
2. Click on Reports in Bottom Navigation
3. Wait until some download is finished and a list is loaded
4. Click  on first report i.e. 'Anc'
5. Do not change anything and click Generate
6. Crashes

**Expected behavior**
Should not crash. Show toast/alert for missing selection or variable needed

### PR DESCRIPTION
**Pull request by AbdulWahabMemon**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #898  

**Type**
Choose one: Bug fix  

**Checklist**
- [x] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [x] I have added any strings visible on UI components to the `strings.xml` file
- [x] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [x] I have built and run the fhircore app to verify my change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportHomeActivity.kt [hunk 1]
```diff
@@ -399,12 +399,17 @@ class ReportHomeActivity : BaseMultiLanguageActivity() {
...
-
-    handleMeasureEvaluate()
+    if (auxPatientData.isNotEmpty()) {
+      val testData = libraryEvaluator.processCQLPatientBundle(auxPatientData)
+      val patientDataStream: InputStream = ByteArrayInputStream(testData!!.toByteArray())
+      patientDataIBase = parser.parseResource(patientDataStream) as IBaseBundle
+      patientResourcesIBase.add(patientDataIBase)
+      handleMeasureEvaluate()
+    } else {
+      var resultItem = ResultItem("Failed", false, "", "Failed", "0")
```
Comment (owais-vd) [inline android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportHomeActivity.kt:409]: hard-coded strings could get from string.xml

File: android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportViewModel.kt [hunk 2]
```diff
@@ -273,9 +278,14 @@ constructor(
...
-      val auxCQLPatientData =
-        parser.encodeResourceToString(fhirResourceDataSource.loadData(patientURL))
-      patientData.postValue(auxCQLPatientData)
+      try {
+        val dataFromUrl = fhirResourceDataSource.loadData(patientURL)
+        val auxCQLPatientData = parser.encodeResourceToString(dataFromUrl)
+        patientData.postValue(auxCQLPatientData)
+      } catch (e: Exception) {
+        e.printStackTrace()
+        patientData.postValue("")
```
Comment (owais-vd) [inline android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportViewModel.kt:287]: Exception should be logged using Timber

**Comment by owais-vd** [comment]
@AbdulWahabMemon needs to add the test cases.

File: android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportHomeActivity.kt [hunk 3]
```diff
@@ -399,12 +399,17 @@ class ReportHomeActivity : BaseMultiLanguageActivity() {
...
   fun handleCQLMeasureLoadPatient(auxPatientData: String) {
-    val testData = libraryEvaluator.processCQLPatientBundle(auxPatientData)
-    val patientDataStream: InputStream = ByteArrayInputStream(testData!!.toByteArray())
-    patientDataIBase = parser.parseResource(patientDataStream) as IBaseBundle
-    patientResourcesIBase.add(patientDataIBase)
-
-    handleMeasureEvaluate()
+    if (auxPatientData.isNotEmpty()) {
+      val testData = libraryEvaluator.processCQLPatientBundle(auxPatientData)
+      val patientDataStream: InputStream = ByteArrayInputStream(testData!!.toByteArray())
```
[the remark this answers is not in the document]
Comment (AbdulWahabMemon) [reply android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportHomeActivity.kt]: updated

File: android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportHomeActivity.kt (hunk 1 above)
Comment (AbdulWahabMemon) [reply android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportHomeActivity.kt]: updated

File: android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportViewModel.kt (hunk 2 above)
Comment (AbdulWahabMemon) [reply android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportViewModel.kt]: updated

File: android/quest/src/test/java/org/smartregister/fhircore/quest/ui/patient/details/QuestPatientDetailScreenTest.kt [hunk 4]
```diff
@@ -153,7 +153,7 @@ class QuestPatientDetailScreenTest : RobolectricTest() {
       .onNodeWithTag(PATIENT_NAME)
       .assertExists()
       .assertIsDisplayed()
-      .assertTextEquals("John Doe, M, 21y")
+      .assertTextEquals("John Doe, M, 22y")
```
Comment (pld) [inline android/quest/src/test/java/org/smartregister/fhircore/quest/ui/patient/details/QuestPatientDetailScreenTest.kt:156]: why is this changed?

File: android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportScreens.kt [hunk 5]
```diff
@@ -273,15 +273,17 @@ fun PatientSelectionBox(
...
-          },
-          onChangeClickListener = {
-            onPatientSelectionChange(ReportViewModel.PatientSelectionType.INDIVIDUAL)
-          }
-        )
+        selectedPatient?.let {
+          SelectedPatientItem(
+            selectedPatient = selectedPatient,
+            onCancelSelectedPatient = {
+              onPatientSelectionChange(ReportViewModel.PatientSelectionType.ALL)
```
Comment (pld) [inline android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportScreens.kt:280]: Can we add a test to cover this line?

File: android/quest/src/test/java/org/smartregister/fhircore/quest/ui/patient/details/QuestPatientDetailScreenTest.kt (hunk 4 above)
Comment (AbdulWahabMemon) [reply android/quest/src/test/java/org/smartregister/fhircore/quest/ui/patient/details/QuestPatientDetailScreenTest.kt:156]: Its being failing in PR,

Comment (pld) [reply android/quest/src/test/java/org/smartregister/fhircore/quest/ui/patient/details/QuestPatientDetailScreenTest.kt]: do you know why it's failing? is there a birthday that's set to December 27th or something like that?

File: android/anc/src/test/java/org/smartregister/fhircore/anc/ui/report/ReportScreensTest.kt [hunk 6]
```diff
@@ -141,6 +143,20 @@ class ReportScreensTest : RobolectricTest() {
     composeRule.onNodeWithTag(REPORT_PATIENT_ITEM).assertExists()
   }
 
+  @Test
+  fun testPatientSelectionChangeListener() {
```
Comment (AbdulWahabMemon) [inline android/anc/src/test/java/org/smartregister/fhircore/anc/ui/report/ReportScreensTest.kt:147]: @pld

File: android/quest/src/test/java/org/smartregister/fhircore/quest/ui/patient/details/QuestPatientDetailScreenTest.kt (hunk 4 above)
Comment (pld) [reply android/quest/src/test/java/org/smartregister/fhircore/quest/ui/patient/details/QuestPatientDetailScreenTest.kt]: I'm not suggesting we should revert this change, I'm trying to understand if this is a brittle test that will fail every time the year changes.

If that's true, we'd want to be able to load the date being used and do some math to convert that to the number of years used here

File: android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportScreens.kt (hunk 5 above)
Comment (AbdulWahabMemon) [reply android/anc/src/main/java/org/smartregister/fhircore/anc/ui/report/ReportScreens.kt:280]: expected test is written in ReportScreenTest class with name testPatientSelectionChangeListener. tagged you there.
this line would be cover there, will re-check if still not coverings with updating params for this fun

**Comment by pld** [comment]
yep I see it, looks great

**Review by pld — DISMISSED** [review]

**Review by pld — APPROVED** [review]

**Pull request merged**